### PR TITLE
Simplify bound params (Part 1/2)

### LINKED ISSUE #35: Bounded track parameters store also free position and momentum
[GitLab issue: #660](https://gitlab.cern.ch/acts/acts-core/issues/660)

The bounded track parameters in `Acts/EventData/SingleTrackParameters.hpp` store 6-vector of local track parameters and derived position and momentum vectors. The later are pure duplicated data and effectively double the size. This duplicate information should be removed. In cases where really the free parameter information is needed, the new free parameters should be used instead.

This originated from the follwoing [discussion](https://gitlab.cern.ch/acts/acts-core/merge_requests/651#note_2931884)

### PR DESCRIPTION
**Pull request by msmk0**
- Implement `SingleBoundTrackParameters` without the `SingleTrackParameters` base.
- Implement `SingleCurvilinearTrackParameters` as a sub-class of the bound parameters. Curvilinear parameters are just bound parameters with a fixed surface type. Consequently, they are implemented as a subclass that only differs in the constructors. This enables simpler code for cases that previously had to handle bound and curvilinear track parameters separately.
- Make the bound parameters const-classes, i.e. they can not be modified after construction, by removing the `.set()` functionality. This greatly simplifies the implementation; e.g. we do not have to recompute the surface geometry for curvilinear parameters anymore. This was only used in the `RiddersPropagator`, where this functionality can be replaced by the construction of a new parameters object.
- Remove cached global position and momentum. The `.position()` and `.momentum()` accessors compute the global information directly. In that case of the position the default geometry context is used.

Supersedes #298 and closes #35.

### CODE REVIEW

File: Core/include/Acts/Propagator/RiddersPropagator.ipp [hunk 1]
```diff
@@ -175,38 +173,13 @@ Acts::RiddersPropagator<propagator_t>::wiggleDimension(
...
-      default:
-        return {};
-    }
+    // Modify start parameter
+    BoundVector values = startPars.parameters();
+    values[param] += h;
+
+    // Propagate with updated start parameters
+    BoundParameters tp(startPars.referenceSurface().getSharedPtr(), values,
+                       startPars.covariance());
```
Comment (asalzburger) [inline Core/include/Acts/Propagator/RiddersPropagator.ipp:182]: Nice.

File: Examples/Algorithms/Propagation/include/ACTFW/Propagation/PropagationAlgorithm.ipp [hunk 2]
```diff
@@ -169,16 +169,16 @@ ProcessCode PropagationAlgorithm<propagator_t>::execute(
     PropagationOutput pOutput;
     if (charge) {
       // charged extrapolation - with hit recording
-      Acts::BoundParameters startParameters(context.geoContext, std::move(cov),
-                                            std::move(pars), surface);
-      sPosition = startParameters.position();
+      Acts::BoundParameters startParameters(surface, std::move(pars),
+                                            std::move(cov));
+      sPosition = startParameters.position(context.geoContext);
       sMomentum = startParameters.momentum();
```
Comment (asalzburger) [inline Examples/Algorithms/Propagation/include/ACTFW/Propagation/PropagationAlgorithm.ipp:175]: I agree, this makes it really clear: the `Context` object is used when it is needed and not at construction of the `BoundParameters`.

File: Examples/Algorithms/Propagation/include/ACTFW/Propagation/PropagationAlgorithm.ipp [hunk 3]
```diff
@@ -169,16 +169,16 @@ ProcessCode PropagationAlgorithm<propagator_t>::execute(
...
       pOutput = executeTest(context, startParameters);
     } else {
       // execute the test for neeutral particles
       Acts::NeutralBoundTrackParameters neutralParameters(
-          context.geoContext, std::move(cov), std::move(pars), surface);
-      sPosition = neutralParameters.position();
+          surface, std::move(pars), std::move(cov));
```
Comment (asalzburger) [inline Examples/Algorithms/Propagation/include/ACTFW/Propagation/PropagationAlgorithm.ipp:180]: I like the new signature:
bound - to where ? `Surface`, what `Parameters`, how `Covariance`.

Very clear.

**Comment by asalzburger** [comment]
Waiting for build problems to be fixed.

**Comment by msmk0** [comment]
@asalzburger I rebased to master, updated to recent changes, and removed the WIP label. After merging #373 the propagation tests should run through. This is now ready to re-reviewed and merged.

File: Tests/IntegrationTests/PropagationTests.hpp [hunk 4]
```diff
@@ -117,9 +117,10 @@ inline Acts::NeutralCurvilinearTrackParameters makeParametersCurvilinearNeutral(
 /// \warning Does not check that they are defined on the same surface.
 template <typename charge_t>
 inline void checkParametersConsistency(
-    const Acts::SingleTrackParameters<charge_t>& cmp,
-    const Acts::SingleTrackParameters<charge_t>& ref, double epsPos,
-    double epsDir, double epsMom) {
+    const Acts::SingleBoundTrackParameters<charge_t>& cmp,
+    const Acts::SingleBoundTrackParameters<charge_t>& ref,
+    const Acts::GeometryContext& geoCtx, double epsPos, double epsDir,
```
Comment (asalzburger) [inline Tests/IntegrationTests/PropagationTests.hpp:122]: Usually we have GeometryContext as first argument, but this is a test file, so no problem.

File: Tests/UnitTests/Core/EventData/TransformFreeToBoundTests.cpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    l0, l1, time, phiInput, theta, qOverP) {
+  // phi is ill-defined in forward/backward tracks
+  auto phi = ((0 < theta) and (theta < M_PI)) ? phiInput : 0.0;
+
+  GeometryContext geoCtx;
+  Vector2D loc(l0, l1);
+  Vector3D pos = Vector3D::Zero();
+  Vector3D dir = makeDirectionUnitFromPhiTheta(phi, theta);
+  // transform reference position
+  surface->localToGlobal(geoCtx, loc, dir, pos);
```
Comment (asalzburger) [inline Tests/UnitTests/Core/EventData/TransformFreeToBoundTests.cpp:41]: We might discuss changing this interface at some point ...

**Review by asalzburger — APPROVED** [review]

File: Tests/IntegrationTests/PropagationTests.hpp (hunk 4 above)
Comment (asalzburger) [reply Tests/IntegrationTests/PropagationTests.hpp:122]: Unrelated to this MR.

File: Tests/UnitTests/Core/EventData/TransformFreeToBoundTests.cpp (hunk 5 above)
Comment (asalzburger) [reply Tests/UnitTests/Core/EventData/TransformFreeToBoundTests.cpp:41]: Unrelated to this MR.

**Review by asalzburger — APPROVED** [review]

**Pull request merged**